File: benchmark/count_prime/count_prime.py
#!/usr/bin/env python3
# Prime counting benchmark for Python
# Counts primes up to LIMIT using trial division
# Reference: pi(10,000,000) = 664,579
#
# How to run:
#   mise run benchmark-count-prime
#
# Or manually:
#   python3 benchmark/count_prime.py

import time


def is_prime(n: int) -> bool:
    if n < 2:
        return False
    d = 2
    while d * d <= n:
        if n % d == 0:
            return False
        d += 1
    return True


def count_primes(limit: int) -> int:
    count = 0
    for n in range(2, limit + 1):
        if is_prime(n):
            count += 1
    return count


def main():
    limit = 10000000

    start = time.perf_counter()
    count = count_primes(limit)
    elapsed = (time.perf_counter() - start) * 1000

    print(f"Prime count up to {limit}: {count}")
    print(f"Elapsed: {elapsed:.2f} ms")


if __name__ == "__main__":
    main()

File: benchmark/count_prime/count_prime.rb
#!/usr/bin/env ruby
# Prime counting benchmark for Ruby
# Counts primes up to LIMIT using trial division
# Reference: π(10,000,000) = 664,579
#
# How to run:
#   mise run benchmark-count-prime
#
# Or manually:
#   ruby benchmark/count_prime.rb

def prime?(n)
  return false if n < 2

  d = 2
  while d * d <= n
    return false if n % d == 0
    d += 1
  end
  true
end

def count_primes(limit)
  count = 0
  (2..limit).each do |n|
    count += 1 if prime?(n)
  end
  count
end

def main
  limit = 10_000_000

  start = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond)
  count = count_primes(limit)
  elapsed_ns = Process.clock_gettime(Process::CLOCK_MONOTONIC, :nanosecond) - start
  elapsed_ms = elapsed_ns / 1_000_000

  puts "Prime count up to #{limit}: #{count}"
  puts "Elapsed: #{elapsed_ms} ms"
end

main
